src/text/writer.rs

```diff
@@ -0,0 +1,352 @@
+            return illegal_operation(format!("Values inside a struct must have a field name."));
+        }
+        if !self.annotations.is_empty() {
+            for annotation in &self.annotations {
+                write!(self.output, "'{}'::", annotation)?;
+            }
+            self.annotations.clear();
+        }
+        Ok(())
+    }
+
+    // Writes:
+    // * the field name (if any)
+    // * the annotations (if any)
+    // * the value written by the `scalar_writer` closure
+    // * a trailing delimiter (if any)
+    fn write_scalar<F>(&mut self, scalar_writer: F) -> IonResult<()>
+    where
+        F: FnOnce(&mut BufWriter<W>) -> IonResult<()>,
+    {
+        self.write_value_metadata()?;
+        scalar_writer(&mut self.output)?;
+        self.write_value_delimiter()?;
+        Ok(())
+    }
+
+    /// Writes an Ion null of the specified type.
+    pub fn write_null(&mut self, ion_type: IonType) -> IonResult<()> {
+        self.write_scalar(|output| TextEncoder::encode_null(output, ion_type))
+    }
+
+    /// Writes the provided bool value as an Ion boolean.
+    pub fn write_bool(&mut self, value: bool) -> IonResult<()> {
+        self.write_scalar(|output| TextEncoder::encode_bool(output, value))
+    }
+
+    /// Writes the provided i64 value as an Ion integer.
+    pub fn write_i64(&mut self, value: i64) -> IonResult<()> {
+        self.write_scalar(|output| TextEncoder::encode_i64(output, value))
+    }
+
+    /// Writes the provided f64 value as an Ion float.
+    pub fn write_f64(&mut self, value: f64) -> IonResult<()> {
+        self.write_scalar(|output| TextEncoder::encode_f64(output, value))
+    }
+
+    /// Writes the provided BigDecimal value as an Ion decimal.
+    pub fn write_big_decimal(&mut self, value: &BigDecimal) -> IonResult<()> {
+        self.write_scalar(|output| TextEncoder::encode_big_decimal(output, value))
+    }
+
+    /// Writes the provided DateTime value as an Ion timestamp.
+    pub fn write_datetime<I: Into<DateTime<FixedOffset>>>(&mut self, value: I) -> IonResult<()> {
+        self.write_scalar(|output| TextEncoder::encode_datetime(output, value))
+    }
+
+    /// Writes the provided &str value as an Ion symbol.
+    pub fn write_str_as_symbol(&mut self, value: &str) -> IonResult<()> {
+        self.write_scalar(|output| TextEncoder::encode_str_as_symbol(output, value))
+    }
+
+    /// Writes the provided &str value as an Ion string.
+    pub fn write_str_as_string(&mut self, value: &str) -> IonResult<()> {
+        self.write_scalar(|output| TextEncoder::encode_str_as_string(output, value))
+    }
+
+    /// Writes the provided byte array slice as an Ion blob.
+    pub fn write_blob(&mut self, value: &[u8]) -> IonResult<()> {
+        self.write_scalar(|output| TextEncoder::encode_blob(output, value))
+    }
+
+    /// Writes the provided byte array slice as an Ion clob.
+    pub fn write_clob(&mut self, _value: &[u8]) -> IonResult<()> {
+        todo!()
```

Comment:
Technically you don't really need to `todo!()` here, can't you just delegate downward to `TextEncoder` (which has the `todo!()`)?

Reply:
Yep, good call.